# End-to-end Fuzzy Deduplication

GPU accelerated implementation of a MinHash-LSH based fuzzy deduplication. For more information about Fuzzy deduplication in NeMo Curator, refer to the [Deduplication](https://docs.nvidia.com/nemo/curator/latest/curate-text/process-data/deduplication/index.html) section of the documentation page.

The tutorial here shows how to run Fuzzy Duplication on text data by executing a 2 end to end workflows which does the following:

1. Read original dataset
2. Compute MinHashes signatures of these documents
3. Perform LSH - Group Minhashes into bands/buckets and shuffle these bands/buckets so that documents in the same bucket are in the same batch/file.
4. Convert the LSH outputs (bucket_id -> doc_id mapping) into a edgelist in preperation for connected components. 
5. Compute connected components across all potential duplicates found via LSH.
6. Generate list of duplicate documents by randomly selecting 1 document to keep from each group/component and dropping the rest.
7. Remove duplicates based on the generated duplicate list.

We also allow users to also run these steps independently, which will be covered in the step by step tutorial in the same directory as this tutorial.

In [1]:
import os

# Silence Curator logs via Loguru
os.environ["LOGURU_LEVEL"] = "ERROR"

import pandas as pd

input_dataset_path = os.path.abspath("./input")  # Path to input dataset
fuzzy_output_dir = os.path.abspath("./output/e2e")  # Path to store all fuzzy outputs including cache & deduped dataset
fuzzy_cache_path = os.path.join(
    fuzzy_output_dir, "cache"
)  # Path to store fuzzy deduplication intermediates (minhash, lsh etc.)
deduplicated_output_path = os.path.join(fuzzy_output_dir, "fuzzy_deduped_dataset")

input_filetype = (
    "parquet"  # this can be either of jsonl or parquet (you'll need to change how input data is generated)
)
output_filetype = "parquet"  # this can be either of jsonl or parquet

In [2]:
from nemo_curator.utils.file_utils import get_all_file_paths_under

if len(get_all_file_paths_under(input_dataset_path)) == 0:
    import os
    import uuid

    from datasets import load_dataset

    input_df = load_dataset("roneneldan/TinyStories", split="train").to_pandas()
    num_rows_per_file = 10_000

    os.makedirs(input_dataset_path, exist_ok=True)

    for i, start_idx in enumerate(range(0, len(input_df), num_rows_per_file)):
        end_idx = min(len(input_df), start_idx + num_rows_per_file)
        subset_df = input_df.iloc[start_idx:end_idx].copy()
        subset_df["id"] = [str(uuid.uuid4()) for _ in range(len(subset_df))]
        subset_df.to_parquet(os.path.join(input_dataset_path, f"part_{i}.parquet"), index=False)

    print(f"Created {len(os.listdir(input_dataset_path))} files")

## Running as a Single Stage (End-to-End)

See the [API Reference](https://docs.nvidia.com/nemo/curator/latest/apidocs/stages/stages.deduplication.fuzzy.workflow.html#api) for more information about the `FuzzyDeduplicationWorkflow` class.

### General Notes
#### ID Generation
1. The Fuzzy Deduplication Workflow doesn't utilize any existing IDs in the input dataset and instead generates IDs on the fly using an ID Generator actor.
2. The ID Generator gives each row a unique increasing integer ID, based on the order files are read.
3. This avoids expensive ID->Integer encoding for the underlying connected components algorithm which only supports integer IDs.
4. When we find duplicates, we save these integer IDs in sorted files with multiple row groups.
5. We also save a `fuzzy_id_generator.json` which maintains a mapping of input file partitions to ID ranges for that batch.
6. During removal, reading the same file groups will give the same integer IDs, using the min/max ID values, we can find all corresponding duplicates in that range making the process faster.

#### Performance Considerations
1. LSH - Configuring bands_per_iteration controls how many bands to process simultanesouly in a single shuffle. Higher values can lead to faster performance but might increase memory pressure.
2. A low `input_blocksize` may not saturate the GPUs enough while a high `input_blocksize` can lead to OOM errors during MinHash and excessive object store usage during removal. It's recommend to keep it at 1-1.5GiB and reduce if running into OOMs during MinHash.
3. The removal step can be memory intensive and it's recommend to set a higher fraction of object store memory for removal (if the machine has enough RAM). The `RayDataExecutor` showed better results during duplicate removal.
4. The removal workflow is CPU only and can be run  on machines that don't have GPUs

#### Hyperparameter Considerations
1. The current defaults for Fuzzy deduplication (260 hashes, 13 hashes per band) approximate finding documents with a jaccard similarity of 0.8. For more information on selecting the number of bands/hashes it's recommended to analyze the S curve and tolerable threshold for false postives (and negatives). More information about LSH can be found in section `3.4.2` [here](http://infolab.stanford.edu/~ullman/mmds/ch3n.pdf).
2. The `char_ngrams` values of 24 is set to approximate roughly ngrams that correspond to ~5 words.


In [3]:
from nemo_curator.stages.deduplication.fuzzy import FuzzyDeduplicationWorkflow
from nemo_curator.stages.deduplication.id_generator import CURATOR_DEDUP_ID_STR
from nemo_curator.stages.text.deduplication import TextDuplicatesRemovalWorkflow

identification_workflow = FuzzyDeduplicationWorkflow(
    cache_path=fuzzy_cache_path,
    output_path=fuzzy_output_dir,
    input_path=input_dataset_path,
    input_filetype=input_filetype,
    input_blocksize="1GiB",
    text_field="text",
    seed=42,
    char_ngrams=24,
    minhashes_per_band=13,
    bands_per_iteration=10,
)

removal_workflow = TextDuplicatesRemovalWorkflow(
    input_path=input_dataset_path,  # Must be identical to the path used during identification
    ids_to_remove_path=os.path.join(fuzzy_output_dir, "FuzzyDuplicateIds"),
    output_path=deduplicated_output_path,
    input_filetype=input_filetype,
    input_blocksize="1GiB",  # This must be identical to the blocksize used during identification
    ids_to_remove_duplicate_id_field=CURATOR_DEDUP_ID_STR,
    id_generator_path=os.path.join(fuzzy_output_dir, "fuzzy_id_generator.json"),
    output_filetype="parquet",
)

In [4]:
from nemo_curator.backends.experimental.ray_data import RayDataExecutor
from nemo_curator.core.client import RayClient

client = RayClient(num_cpus=64, num_gpus=2)  # change as needed
client.start()

_ = identification_workflow.run()
_ = removal_workflow.run(executor=RayDataExecutor())

2025-11-20 09:13:20,334	INFO worker.py:1692 -- Using address 127.0.1.1:6382 set in the environment variable RAY_ADDRESS
2025-11-20 09:13:20,336	INFO worker.py:1833 -- Connecting to existing Ray cluster at address: 127.0.1.1:6382...
[2025-11-20 09:13:25,485 W 2843400 2843400] global_state_accessor.cc:505: Some processes that the driver needs to connect to have not registered with GCS, so retrying. Have you run 'ray start' on this node?
[2025-11-20 09:13:26,486 W 2843400 2843400] global_state_accessor.cc:505: Some processes that the driver needs to connect to have not registered with GCS, so retrying. Have you run 'ray start' on this node?
[2025-11-20 09:13:27,487 W 2843400 2843400] global_state_accessor.cc:505: Some processes that the driver needs to connect to have not registered with GCS, so retrying. Have you run 'ray start' on this node?
2025-11-20 09:13:27,613	INFO utils.py:87 -- Overwriting previous Ray address (127.0.1.1:6381). Running ray.init() on this node will now connect to 

2025-11-20 09:13:22,775	INFO usage_lib.py:447 -- Usage stats collection is disabled.
2025-11-20 09:13:22,776	INFO scripts.py:914 -- Local node IP: 127.0.1.1
2025-11-20 09:13:27,612	SUCC scripts.py:950 -- --------------------
2025-11-20 09:13:27,612	SUCC scripts.py:951 -- Ray runtime started.
2025-11-20 09:13:27,612	SUCC scripts.py:952 -- --------------------
2025-11-20 09:13:27,612	INFO scripts.py:954 -- Next steps
2025-11-20 09:13:27,612	INFO scripts.py:957 -- To add another node to this Ray cluster, run
2025-11-20 09:13:27,612	INFO scripts.py:960 --   ray start --address='127.0.1.1:6382'
2025-11-20 09:13:27,613	INFO scripts.py:969 -- To connect to this Ray cluster:
2025-11-20 09:13:27,613	INFO scripts.py:971 -- import ray
2025-11-20 09:13:27,613	INFO scripts.py:972 -- ray.init(_node_ip_address='127.0.1.1')
2025-11-20 09:13:27,613	INFO scripts.py:984 -- To submit a Ray job using the Ray Jobs CLI:
2025-11-20 09:13:27,613	INFO scripts.py:985 --   RAY_API_SERVER_ADDRESS='http://127.0.0.1

[2025-11-20 09:13:28,488 I 2843400 2843400] global_state_accessor.cc:487: This node has an IP address of 127.0.1.1, but we cannot find a local Raylet with the same address. This can happen when you connect to the Ray cluster with a different IP address or when connecting to a container.
2025-11-20 09:13:28,491	INFO worker.py:2013 -- Connected to Ray cluster.
/raid/adattagupta/NeMo-Curator/.venv/lib/python3.12/site-packages/ray/_private/worker.py:2052: FutureWarning: Tip: In future versions of Ray, Ray will no longer override accelerator visible devices env var if num_gpus=0 or num_gpus=None (default). To enable this behavior and turn off this error message, set RAY_ACCEL_ENV_VAR_OVERRIDE_ON_ZERO=0
  warnings.warn(
2025-11-20 09:13:32,986	INFO worker.py:1692 -- Using address 127.0.1.1:6382 set in the environment variable RAY_ADDRESS
2025-11-20 09:13:32,988	INFO worker.py:1833 -- Connecting to existing Ray cluster at address: 127.0.1.1:6382...
2025-11-20 09:13:32,994	INFO worker.py:2004 

Running 0: 0.00 row [00:00, ? row/s]

[2025-11-20 09:15:13,792 E 2843400 2843400] core_worker.cc:2153: Actor with class name: 'MapWorker(MapBatches(ParquetReaderStageActor))' and ID: 'c956838f83654fc601f9e42f06000000' has constructor arguments in the object store and max_restarts > 0. If the arguments in the object store go out of scope or are lost, the actor restart will fail. See https://github.com/ray-project/ray/issues/53727 for more details.


- MapBatches(FilePartitioningStageTask) 1: 0.00 row [00:00, ? row/s]

- StreamingRepartition 2: 0.00 row [00:00, ? row/s]

- MapBatches(ParquetReaderStageActor) 3: 0.00 row [00:00, ? row/s]

- MapBatches(TextDuplicatesRemovalStageTask)->MapBatches(ParquetWriterTask) 4: 0.00 row [00:00, ? row/s]

2025-11-20 09:15:13,819	WARNING resource_manager.py:134 -- ⚠️  Ray's object store is configured to use only 12.0% of available memory (186.3GiB out of 1555.2GiB total). For optimal Ray Data performance, we recommend setting the object store to at least 50% of available memory. You can do this by setting the 'object_store_memory' parameter when calling ray.init() or by setting the RAY_DEFAULT_OBJECT_STORE_MEMORY_PROPORTION environment variable.
2025-11-20 09:15:42,543	INFO streaming_executor.py:279 -- ✔️  Dataset dataset_5_0 execution finished in 28.79 seconds
2025-11-20 09:15:42,592	INFO worker.py:1692 -- Using address 127.0.1.1:6382 set in the environment variable RAY_ADDRESS
2025-11-20 09:15:42,593	INFO worker.py:1833 -- Connecting to existing Ray cluster at address: 127.0.1.1:6382...
2025-11-20 09:15:42,599	INFO worker.py:2004 -- Connected to Ray cluster. View the dashboard at http://127.0.0.1:8268 


### Looking at Intermediate Results and Output

#### MinHash Results
1. `_curator_dedup_id` - The IDs assigned to this dataset on the fly during the intial read.
2. `_minhash_signature` - MinHash Signature

#### LSH Results
1. `_bucket_id` - The bucket/band identifier
2. `_curator_dedup_id` - List of all document IDs that belong to that bucket

#### Buckets To Edges Result
1. `_curator_dedup_id_x`, `_curator_dedup_id_y` - Mapping of edges in a Graph where each column are documents that are potential duplicates.

In [5]:
minhash_path = os.path.join(fuzzy_cache_path, "MinHashStage")
display(pd.read_parquet(os.path.join(minhash_path, os.listdir(minhash_path)[0])).head())

lsh_path = os.path.join(fuzzy_cache_path, "LSHStage")
display(pd.read_parquet(os.path.join(lsh_path, os.listdir(lsh_path)[0])).head())

b2e_path = os.path.join(fuzzy_cache_path, "BucketsToEdgesStage")
display(pd.read_parquet(os.path.join(b2e_path, os.listdir(b2e_path)[0])).head())

_curator_dedup_id                                 _minhash_signature
0                  0  [11644717, 429172, 6014805, 86354, 2387151, 49...
1                  1  [2103321, 653305, 2941429, 5780991, 6977799, 7...
2                  2  [1891498, 3797631, 2961751, 50078, 21382505, 5...
3                  3  [1286357, 4060996, 1376561, 3044837, 7369355, ...
4                  4  [6272013, 12535265, 819579, 5975720, 25677928,...

_bucket_id   _curator_dedup_id
0  b0_000055fd7daae1e46223e8b7e06bf2e0  [1158375, 2079489]
1  b0_00006a5f30f7b2c96588bfc1bfb5321a   [365218, 1933514]
2  b0_00006f316d5bd251bd83702e3f1e017f    [161590, 771961]
3  b0_0000d90e9e4140a7ac31e6b227a62f62      [8290, 567169]
4  b0_0000f975e5bcda25838df43b0d37737f   [965853, 1334885]

_curator_dedup_id_x  _curator_dedup_id_y
0               150383               921559
1               679249              1126071
2               128422              1370422
3                29885               516200
4               270728              1089921

#### Connected Components Result

1. `_curator_dedup_id` - The document IDs
2. `_duplicate_group_id` - The group ID that document belongs to. Documents with the same duplicate group ID are duplicates

In [6]:
cc_path = os.path.join(fuzzy_cache_path, "ConnectedComponentsStage")
cc_df = pd.read_parquet(cc_path)  # works with pandas since the input here is small
display(cc_df)
grouped_cc_df = cc_df.groupby("_duplicate_group_id")._curator_dedup_id.agg(list)
display(grouped_cc_df)
duplicate_cluster_sizes = cc_df._duplicate_group_id.value_counts()
display(duplicate_cluster_sizes)

_curator_dedup_id  _duplicate_group_id
0                     576                    0
1                     577               482206
2                     578                    2
3                     579               482207
4                     581               482209
...                   ...                  ...
640509            2119669               640509
640510            2119670               640510
640511            2119671               479940
640512            2119673               479941
640513            2119675               640513

[640514 rows x 2 columns]

_duplicate_group_id
0              [576, 197440]
2              [578, 197442]
5              [583, 197447]
6              [584, 197448]
7              [585, 197449]
                 ...        
640507    [1942715, 2119666]
640508    [1942716, 2119667]
640509    [1942718, 2119669]
640510    [1942719, 2119670]
640513    [1942724, 2119675]
Name: _curator_dedup_id, Length: 320043, dtype: object

_duplicate_group_id
463522    230
504398      3
88059       3
88060       3
559408      3
         ... 
551514      2
106803      2
514494      2
514493      2
480738      2
Name: count, Length: 320043, dtype: int64

Based on the distribution above we can see that there is one cluster/group where 230 documents are all duplicates followed by many smaller clusters with 2/3 documents that are duplicates.

#### FuzzyDuplicateIds Results (List of duplicate docs to remove)
1. `_curator_dedup_id` - ID of docs in the removal list

In [7]:
duplicate_ids_path = os.path.join(fuzzy_output_dir, "FuzzyDuplicateIds")
duplicates_df = pd.read_parquet(duplicate_ids_path)
display(duplicates_df.head())

print(f"Number of duplicate documents found for removal: {len(duplicates_df)}")

_curator_dedup_id
0                577
1                579
2                581
3                584
4                585

Number of duplicate documents found for removal: 320471


#### Checking that the duplicate ids list contains only one document per group

In [8]:
# As an example let's look at the group with the largest number of duplicates
largest_duplicate_cluster = grouped_cc_df.loc[duplicate_cluster_sizes.index[0]]

# number of docs in the removal list from this group
docs_to_remove_in_group = duplicates_df._curator_dedup_id.isin(largest_duplicate_cluster).sum()

print(f"Number of documents in the duplicate group: {len(largest_duplicate_cluster)}")
print(f"Number of documents in the removal list from the same group: {docs_to_remove_in_group}")
assert docs_to_remove_in_group == (len(largest_duplicate_cluster) - 1)  # noqa: S101

Number of documents in the duplicate group: 230
Number of documents in the removal list from the same group: 229


#### Advanced: Looking at examples of duplicate documents

1. This analysis involves re-reading the input data with the same ID mapping that was used during duplicate identification.
2. Merging the input data with the connected components results on the `_curator_dedup_id` column to associate each document which the duplicate group it belongs to which can be used for further analysis.

**NOTE**: This analsis approach is itended as an example for smaller datasets and only works for cases where the connected components dataframe is small and fits comfortable in memory. It is not recommended for larger datasets.

In [9]:
from nemo_curator.pipeline import Pipeline
from nemo_curator.stages.base import ProcessingStage
from nemo_curator.stages.resources import Resources
from nemo_curator.stages.text.io.reader import ParquetReader
from nemo_curator.tasks.document import DocumentBatch


class CustomMergeStage(ProcessingStage[DocumentBatch, DocumentBatch]):
    """
    Warning: This should not be attempted with large connected components results.
    A small stage that merges the input data (using the id's generated) with the connected components result.
    Works because CC results are small enough to fit per batch.
    """

    _resources = Resources(cpus=1.0)

    def process(self, batch: DocumentBatch) -> DocumentBatch:
        df = batch.to_pandas().merge(cc_df, how="inner", on=[CURATOR_DEDUP_ID_STR])
        return DocumentBatch(
            task_id=batch.task_id, dataset_name=batch.dataset_name, data=df, _stage_perf=batch._stage_perf
        )


pipeline = Pipeline(
    name="Explore duplicates",
    stages=[ParquetReader(file_paths=input_dataset_path, blocksize="1GiB", _assign_ids=True), CustomMergeStage()],
)

In [10]:
from nemo_curator.core.client import RayClient
from nemo_curator.stages.deduplication.id_generator import create_id_generator_actor, kill_id_generator_actor

client = RayClient(num_cpus=8)  # change as needed
client.start()

create_id_generator_actor(filepath=os.path.join(fuzzy_output_dir, "fuzzy_id_generator.json"))
merged_results = pipeline.run()
merged_df = pd.concat([batch.to_pandas() for batch in merged_results]).sort_values("_duplicate_group_id")
kill_id_generator_actor()

2025-11-20 09:22:33,038	INFO worker.py:1692 -- Using address 127.0.1.1:6382 set in the environment variable RAY_ADDRESS
2025-11-20 09:22:33,039	INFO worker.py:1833 -- Connecting to existing Ray cluster at address: 127.0.1.1:6382...
2025-11-20 09:22:33,039	INFO worker.py:1851 -- Calling ray.init() again after it has already been called.
2025-11-20 09:22:33,745	INFO worker.py:1692 -- Using address 127.0.1.1:6382 set in the environment variable RAY_ADDRESS
2025-11-20 09:22:33,747	INFO worker.py:1833 -- Connecting to existing Ray cluster at address: 127.0.1.1:6382...
2025-11-20 09:22:33,752	INFO worker.py:2004 -- Connected to Ray cluster. View the dashboard at http://127.0.0.1:8268 
2025-11-20 09:22:33,768	INFO worker.py:1692 -- Using address 127.0.1.1:6382 set in the environment variable RAY_ADDRESS
2025-11-20 09:22:33,769	INFO worker.py:1833 -- Connecting to existing Ray cluster at address: 127.0.1.1:6382...
2025-11-20 09:22:33,769	INFO worker.py:1851 -- Calling ray.init() again after it

In [11]:
display(merged_df[merged_df._curator_dedup_id.isin(largest_duplicate_cluster)])

text                                    id  _curator_dedup_id  \
369299       b8f7e80f-5c03-4686-8502-36c93874fe23            1215014   
93041        1c2dfd9b-7f03-468d-9d1c-d53848129834             304132   
258686       d466d372-2468-428a-9b4e-b0db0e545e36             856047   
369307       49eba958-d175-4969-9eee-87ca0b42ab5e            1215084   
557145       f5d0e2d8-b5ba-4996-b44b-2da667e26807            1855005   
...     ...                                   ...                ...   
557151       9e493d0e-c479-4465-9c35-431765d36484            1855012   
369309       96a15122-9f50-41ab-8911-59afee17af60            1215088   
192032       573e97d5-2229-457d-9e41-b98f4ee901b5             630284   
368142       7bc9641f-4e2e-4771-ae40-0f12603a7695            1208884   
369312       2903d9bf-a823-4649-a2d1-278a9e22e8d2            1215095   

        _duplicate_group_id  
369299               463522  
93041                463522  
258686               463522  
369307               463522  
557145               463522  
...                     ...  
557151               463522  
369309               463522  
192032               463522  
368142               463522  
369312               463522  

[230 rows x 4 columns]

The largest cluster/group of duplicates in this dataset seems to be all documents with empty/no text.

Let's look at the second largest cluster of documents.

In [12]:
duplicates = merged_df[merged_df._curator_dedup_id.isin(grouped_cc_df.loc[duplicate_cluster_sizes.index[1]])]
display(duplicates)

print(f"\nDocument1\n----------\n{duplicates.iloc[0].text}")
print(f"\nDocument2\n----------\n{duplicates.iloc[1].text}")

text  \
32338   Once upon a time, there was a little boy named...   
248446  Once upon a time, there was a little boy named...   
170411  Once upon a time, there was a little boy named...   

                                          id  _curator_dedup_id  \
32338   b44cee45-e69d-4c30-b1d8-8ba04996d9a3              99567   
248446  3b551a23-3588-490c-ba03-941ea481e33c             823161   
170411  2a7ccbf2-97da-4bb9-9906-6c019e9af8e6             561485   

        _duplicate_group_id  
32338                504398  
248446               504398  
170411               504398


Document1
----------
Once upon a time, there was a little boy named Timmy. Timmy loved to play with his toy cars and make them go zoom. One day, Timmy saw some big kids playing basketball. Timmy wanted to play too, but the big kids said he was too young. Timmy felt frustrated and sad. 

Timmy went home and told his mom about what happened. His mom said, "Timmy, just because you are young doesn't mean you can't play. You should ask the big kids if you can play with them." 

The next day, Timmy went back to the basketball court and asked the big kids if he could play. They said yes! Timmy was so happy and he played with the big kids all day. But as the sun began to set, Timmy's ball accidentally hit a car and the owner of the car got very angry. Timmy didn't know what to do and he felt scared. 

The moral of the story is that sometimes things don't go the way we want them to, but we should always try our best and be responsible for our actions.

Document2
----------
Once upon a time, th

In [13]:
client.stop()